src/uri.ts

```diff
@@ -7,7 +7,7 @@ import { Endpoint, Settings } from './settings'
 export interface ResolvedURI {
     repo: string
     rev: string
-    path: string
+    path?: string
```

Comment:
Could you document here under which condition this would be `undefined`? The docblock above says the interface identifies a file path, so intuitively not including a `path` does not make sense to me

Reply:
@felixfbecker context:
- In https://github.com/codecov/sourcegraph-codecov/pull/34/files#diff-8a415beef63a505bd17d360a4c9dce13 I changed `resolveURI()` so that the implementation would match the returned type.
- Turns out `resolveURI` can be called on a root URI, not just a file URI, in which case the URI does not contain a path.

Maybe for clarity & optimal type safety we should separate `resolveRootURI` and `resolveFilURI`:


```typescript
interface ResolvedRootURI {
    repo: string
    rev: string
}

function resolveRootURI(uri: string): ResolvedRootURI {}

interface ResolvedFileURI extends ResolvedRootURI {
    path: string
}

function resolveFileURI(uri: string): ResolvedFileURI {}
```

Thoughts @felixfbecker @hadrian-git ?

Reply:
```ts
/**
 * A resolved URI without an identified path.
 */
export interface ResolvedRootURI {
    repo: string
    rev: string
}

/**
 * A resolved URI with an identified path.
 */
export interface ResolvedFileURI extends ResolvedRootURI {
    path: string
}

/**
 * Resolve a URI of the form git://github.com/owner/repo?rev#path and file:///path to an absolute reference
 */
export function resolveFileURI(uri: string): ResolvedFileURI {
    // ...
    return {
        repo,
        rev,
        path,
    }
}

/**
 * Resolve a URI of the form git://github.com/owner/repo?rev to an absolute reference
 */
export function resolveRootURI(uri: string): ResolvedRootURI {
    const { path, ...result } = resolveFileUri(uri)
    return result
}
```
wdyt?

Also, can files still have a URI of form `file:///...`? I saw in the test file `uri.test.ts` that `file` is set as an unsupported scheme. Should I update the tests, docs, e.t.c?

Reply:
I would prefer two functions as one of them can throw an Error. But the most important part is proper documentation (which caused this bug in the first place)

Reply:
yep, can also be

```ts
function resolveURI(url: URL): ResolvedRootURI {
    // ...
    return { repo, rev }
}

export function resolveFileURI(uri: string): ResolvedFileURI {
    const url = new URL(uri)
    const path = url.hash.slice(1)
    return {
        ...resolveURI(url),
        path,
    }
}

export function resolveRootURI(uri: string): ResolvedRootURI {
    return resolveURI(new URL(uri))
}
```